### ISSUE TITLE
Blazor posibility to wait for OnInitAsync before initial rendering

### ISSUE THREAD

**Issue by isc30** [dotnet/aspnetcore#21285]
This is somehow related to #9190.

When using prerendering, the full page is rendered (with all the components rendered **after *OnInitAsync***). When the client side loads, it will **not** wait for *OnInitAsync* before the initial render, making the page "pop-in" for some content that was already there.

Rendering before *OnInitAsync* has finished is a great idea for implementing "loading" screens, but it doesn't play well with the initial render.

Is there any way to avoid that "pop-in" effect in place? If not, how could it be implemented?
Thanks

**Comment by javiercn** [comment]
@isc30 thanks for contacting us.

You can cache the data during the prerendering phase and load it synchronously after that. If you do it that way there is no async code path and you don't get the "pop-in"

**Comment by isc30** [comment]
Hi @javiercn, in this specific case I need to fetch some json from an endpoint, but I don't want to "ruin" the prerendered page by having a "loading" message for 100ms. There is no way to just "keep" what's in the page until I got every component fully initialized.

I can't find any way of doing this nicely at the moment, that's why i was suggesting introducing a method/flag in the WebAssemblyHost to defer the render after every component's *OnInitAsync* has finished.

**Comment by mkArtakMSFT** [comment]
We've moved this issue to the Backlog milestone. This means that it is not going to happen for the coming release. We will reassess the backlog following the current release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by isc30** [comment]
Hi, this could be considered a bug. Imagine the following scenario:

```cs
public class MyComponent : ComponentBase
{
    protected override bool ShouldRender() => false;
}
```

This will trigger the initial render either way, completely ignoring the flag.
Please consider this minimal change before the main release, since it opens the possibility to create custom hydraters for prerendered content.

**Comment by javiercn** [comment]
@isc30 I'm not sure what you are asking us to do,

but changing something on ComponentBase is not possible for us until/after the 5.0 release. This is part of the component model so it can't only be applied to Blazor WebAssembly and it is not an area where we plan to make changes.

We'll look into documenting the pattern to follow here in the future.

**Comment by isc30** [comment]
Similar fix from #22202 would solve this too

**Comment by srpeirce** [comment]
Yes this would make blazor a much more viable option for customer facing sites. At the moment it is a hard sell compared to JS libraries/frameworks because of these limitations. 
We need the server side rendering (SSR) for fast paint times and better SEO, but the reload-flash leads to poor user experience. The slow time to interactive is not ideal in this scenario (whilst blazor loads) but we can atleast pull a few tricks to cover this up/damage control. 
Using wasm without SSR leads to a ~4 second page load on my Galaxy Note 8 which is definitely not viable for customer facing site.

**Comment by mihaimyh** [comment]
No love for SEO from Microsoft then?
